eclair-core/src/main/scala/fr/acinq/eclair/swap/SwapInSender.scala

```diff
@@ -0,0 +1,333 @@
+
+                                            "Refund Cooperatively"
+                                     |            CoopClose           | [sendCoopClose]
+                                     |<-------------------------------|
+     (claim_by_coop) [claimSwapCoop] |                                |
+
+                                             "Refund After Csv"
+                           [waitCsv] |                                |
+                                     |                                |
+       (claim_by_csv) [claimSwapCsv] |                                |
+
+  */
+
+  def apply(nodeParams: NodeParams, watcher: ActorRef[ZmqWatcher.Command], register: actor.ActorRef, wallet: OnChainWallet): Behavior[SwapCommands.SwapCommand] =
+    Behaviors.setup { context =>
+      Behaviors.receiveMessagePartial {
+        case StartSwapInSender(replyTo, amount, swapId, channelId) =>
+          new SwapInSender(replyTo, amount, swapId, channelId, nodeParams, watcher, register, wallet, context)
+            .createSwap()
+        case RestoreSwapInSender(replyTo, d) =>
+          new SwapInSender(replyTo, d.request.amount.sat, d.request.swapId, d.channelId, nodeParams, watcher, register, wallet, context)
+            .awaitOpeningTxConfirmed(d.request, d.agreement, d.invoice, d.openingTxBroadcasted)
+        case AbortSwapInSender => Behaviors.stopped
+      }
+    }
+}
+
+private class SwapInSender(replyTo: ActorRef[SwapResponses.Response], amount: Satoshi, swapId: String, channelId: ByteVector32, nodeParams: NodeParams, watcher: ActorRef[ZmqWatcher.Command], register: actor.ActorRef, wallet: OnChainWallet, implicit val context: ActorContext[SwapCommands.SwapCommand]) {
+  val protocolVersion = 1
+  val noAsset = ""
+  implicit val timeout: Timeout = 30 seconds
+  private val keyManager: SwapKeyManager = nodeParams.swapKeyManager
+  private implicit val feeRatePerKw: FeeratePerKw = nodeParams.onChainFeeConf.feeEstimator.getFeeratePerKw(target = nodeParams.onChainFeeConf.feeTargets.fundingBlockTarget)
+  private val maxPremium = (feeRatePerKw * claimByInvoiceTxWeight / 1000).toLong // TODO: how should swap sender calculate an acceptable premium?
+
+  private def makerPrivkey(): PrivateKey = keyManager.openingPrivateKey(SwapKeyManager.keyPath(swapId)).privateKey
+  private def makerPubkey(): PublicKey = makerPrivkey().publicKey
+  private def takerPubkey(agreement: SwapInAgreement): PublicKey = PublicKey(ByteVector.fromValidHex(agreement.pubkey))
+
+  private def createSwap(): Behavior[SwapCommand] = {
+    // setup conditions satisfied, create the swap
+    val pubKey = keyManager.openingPublicKey(SwapKeyManager.keyPath(swapId)).publicKey.toString()
+    queryOnChainBalance(wallet)
+
+    receiveSwapMessage[CreateSwapMessages](context, "createSwap") {
+      case QueryOnChainBalance(Success(onChainBalance: OnChainBalance)) if onChainBalance.confirmed < amount + maxPremium.sat =>
```

Comment:
I don't know if this check and the `case ChannelInfoResponse(RES_GET_CHANNEL_INFO(_, _, _, d: DATA_NORMAL)) if d.commitments.availableBalanceForReceive.truncateToSatoshi < amount` are really necessary, especially for a first version.

I'd rather stay simple at first and just abort later when we realize we actually cannot pay (either when failing to fund an on-chain transaction or failing to send an HTLC).

We can add nice-to-have checks later, I think we should just focus on an MVP at first, which will help focus on the core of the protocol.

Reply:
Good point; this check is removed in recent versions.